engine.py

```diff
@@ -23,200 +19,240 @@
         import tk_nuke
         
-        # make sure callbacks tracking the context switching are active
+        # Make sure callbacks tracking the context switching are active.
         tk_nuke.tank_ensure_callbacks_registered()
-        
-       
+
     def post_app_init(self):
         """
-        Called when all apps have initialized
+        Called when all apps have initialized.
         """
-        # render the menu!
-        if self._ui_enabled:
-            
-            # note! not using the import as this confuses nuke's calback system
+        # Render the menu!
+        if self.has_ui:
+            # Note! not using the import as this confuses nuke's calback system
             # (several of the key scene callbacks are in the main init file...)            
             import tk_nuke
 
             menu_name = "Shotgun"
             if self.get_setting("use_sgtk_as_menu_name", False):
                 menu_name = "Sgtk"
-            
-            # create menu
+
             self._menu_generator = tk_nuke.MenuGenerator(self, menu_name)
             self._menu_generator.create_menu()
-            
-            # initialize favourite dirs in the file open/file save dialogs
-            self.__setup_favorite_dirs()
-            
-            # register all panels with nuke's callback system
-            # this will be used at nuke startup in order
-            # for nuke to be able to restore panels 
-            # automatically. For all panels that exist as
-            # part of saved layouts, nuke will look through
-            # a global list of registered panels, try to locate
-            # the one it needs and then run the callback         
-            for (panel_id, panel_dict) in self.panels.iteritems():
-                nukescripts.panels.registerPanel(panel_id, panel_dict["callback"])
-            
-        # iterate over all apps, if there is a gizmo folder, add it to nuke path
+
+            if self.hiero:
+                import hiero
+                def set_project_root(event):
+                    """Ensure any new projects get the project root or default startup 
+                    projects get the project root set
+                    """ 
+                    for p in hiero.core.projects():
+                        if not p.projectRoot():
+                            self.log_debug(
+                                "Setting projectRoot on %s to: %s" % (
+                                    p.name(),
+                                    self.tank.project_path
+                                )
+                            )
+                            p.setProjectRoot(self.tank.project_path)
+                hiero.core.events.registerInterest(
+                    'kAfterNewProjectCreated',
+                    set_project_root,
+                )
+                # We don't need to get into the Gizmo stuff after this block
+                # since that is specific to Nuke.
+                return
```

Comment:
i think we should avoid this if possible - it easily missed and may cause confusion in the future if we make further changes.